File: scala/pal/src/main/scala/pal/ScaffoldCircuitInput.scala
package pal

import Expr.TRUE
import Ref.{EMPTY, NEW, PREVIOUS}
import ScaffoldCircuit.*

/** Clonable readonly letter/gap input heads in the symbolic scaffold circuit. */
object ScaffoldCircuitInput {
  type HeadPools = Map[String, StackPool]
  type CounterPools = Map[String, Map[String, StackPool]]

  private def headPools(pool: StackPool): HeadPools = Map("l" -> pool, "r" -> pool)
  private def queuePools(pool: StackPool): HeadPools =
    Vector("F", "B", "B2", "Fr", "Br", "WF", "WB").map(_ -> pool).toMap
  private def counterPools(pool: StackPool): CounterPools =
    Map("m" -> Map("pos" -> pool, "neg" -> pool), "c" -> Map("pos" -> pool, "neg" -> pool))

  final class InputHead(pools: HeadPools, queues: HeadPools, counters: CounterPools,
                        val name: String, val alphabet: String = "ab") {
    def this(headPool: StackPool, queuePool: StackPool, counterPool: StackPool, name: String) = {
      this(headPools(headPool), queuePools(queuePool), counterPools(counterPool), name, "ab")
    }
    val circuit: Circuit = pools("l").circuit
    val focusKey: String = name + ".focus"
    var focus: Ref = circuit.getRef(PREVIOUS, focusKey)
    val leftStack: Stack = new Stack(pools("l"), name + ".l")
    val rightStack: Stack = new Stack(pools("r"), name + ".r")
    val incoming: Queue = new Queue(queues, counters, name + ".in")

    def append(cell: Ref = NEW, enabled: Expr = TRUE): Unit = {
      incoming.push(cell, enabled)
      incoming.work(enabled)
    }

    def read(): Value[Any] = {
      val symbol = circuit.get(focus, "input", alphabet.toVector, alphabet.head)
      Value.select(focus.present(), symbol, Value.constant(None))
    }

    def canRight(): Expr = disjunction(neg(rightStack.empty()), neg(incoming.empty()))

    /** Read the next cell at an operation boundary without moving the head. */
    def peekRight(): Ref = {
      val saved = rightStack.peek()._1
      val arrived = incoming.stacks("F").peek()._1
      val fromStack = neg(rightStack.empty())
      circuit.require(neg(incoming.stacks("F").empty()), conjunction(neg(fromStack), neg(incoming.empty())))
      Ref.select(fromStack, saved, arrived)
    }

    def reset(enabled: Expr = TRUE): Unit = {
      focus = Ref.select(enabled, EMPTY, focus)
      leftStack.clear(enabled)
      rightStack.clear(enabled)
      incoming.clear(enabled)
    }

    def right(enabled: Expr = TRUE): Unit = {
      circuit.require(canRight(), enabled)
      leftStack.push(focus, enabled = enabled)
      val fromStack = neg(rightStack.empty())
      val fromQueue = conjunction(enabled, neg(fromStack))
      val saved = rightStack.pop(conjunction(enabled, fromStack))._1
      incoming.work(fromQueue)
      val arrived = incoming.pop(fromQueue)
      incoming.work(fromQueue)
      focus = Ref.select(enabled, Ref.select(fromStack, saved, arrived), focus)
    }

    def left(enabled: Expr = TRUE): Unit = {
      circuit.require(neg(leftStack.empty()), enabled)
      rightStack.push(focus, enabled = enabled)
      val saved = leftStack.pop(enabled)._1
      focus = Ref.select(enabled, saved, focus)
    }

    def copyFrom(other: InputHead, enabled: Expr = TRUE): Unit = {
      focus = Ref.select(enabled, other.focus, focus)
      leftStack.copyFrom(other.leftStack, enabled)
      rightStack.copyFrom(other.rightStack, enabled)
      incoming.copyFrom(other.incoming, enabled)
    }

    def commit(maintain: Boolean = true): Unit = {
      if (maintain) { incoming.work() }
      leftStack.commit()
      rightStack.commit()
      incoming.commit()
      circuit.putRef(focusKey, focus)
    }
  }

  final class PlaceHead(pools: HeadPools, queues: HeadPools, counters: CounterPools,
                        val name: String, alphabet: String = "ab") {
    def this(headPool: StackPool, queuePool: StackPool, counterPool: StackPool, name: String) = {
      this(headPools(headPool), queuePools(queuePool), counterPools(counterPool), name, "ab")
    }
    val head: InputHead = new InputHead(pools, queues, counters, name, alphabet)
    val circuit: Circuit = head.circuit
    val gapKey: String = name + ".gap"
    var gap: Expr = circuit.get(PREVIOUS, gapKey, Vector(false, true), true).eqTo(true)

    def append(cell: Ref, enabled: Expr = TRUE): Unit = head.append(cell, enabled)

    def read(): Value[Any] = {
      val value = head.read()
      Value.select(value.eqTo(None), Value.constant(None), Value.select(gap, Value.constant('s'), value))
    }

    def canRight(): Expr = disjunction(neg(gap), head.canRight())

    def right(enabled: Expr = TRUE): Unit = {
      head.right(conjunction(enabled, gap))
      gap = choose(enabled, neg(gap), gap)
    }

    def left(enabled: Expr = TRUE): Unit = {
      circuit.require(neg(head.read().eqTo(None)), enabled)
      head.left(conjunction(enabled, neg(gap)))
      gap = choose(enabled, neg(gap), gap)
    }

    def isFirst(): Expr = {
      val before = head.leftStack.peek()._1
      conjunction(neg(gap), neg(head.leftStack.empty()), neg(before.present()))
    }

    def copyFrom(other: PlaceHead, enabled: Expr = TRUE): Unit = {
      head.copyFrom(other.head, enabled)
      gap = choose(enabled, other.gap, gap)
    }

    def commit(): Unit = {
      head.commit()
      circuit.put(gapKey, Value.select(gap, Value.constant(true), Value.constant(false)))
    }
  }
}

File: docs/palindromes-in-peg/scaffold_circuit_input.py
"""Clonable readonly letter/gap input heads in the symbolic scaffold circuit."""
from scaffold_circuit import (Value, Ref, EMPTY, PREVIOUS, TRUE, FALSE,
                              choose, conjunction, disjunction, neg)
from scaffold_circuit_structs import Stack, Queue


class InputHead:
  def __init__(self, head_pool, queue_pool, counter_pool, name, alphabet="ab"):
    pools = head_pool if isinstance(head_pool, dict) else {"l": head_pool, "r": head_pool}
    self.circuit, self.name, self.alphabet = pools["l"].circuit, name, tuple(alphabet)
    self.focus_key = name + ".focus"
    self.focus = self.circuit.get_ref(PREVIOUS, self.focus_key)
    self.left_stack, self.right_stack = Stack(pools["l"], name + ".l"), Stack(pools["r"], name + ".r")
    self.incoming = Queue(queue_pool, counter_pool, name + ".in")

  def append(self, cell, enabled=TRUE):
    self.incoming.push(cell, enabled)
    self.incoming.work(enabled)

  def read(self):
    symbol = self.circuit.get(self.focus, "input", self.alphabet, self.alphabet[0])
    return Value.select(self.focus.present(), symbol, Value.constant(None))

  def can_right(self):
    return disjunction(neg(self.right_stack.empty()), neg(self.incoming.empty()))

  def peek_right(self):
    """Read the next cell at an operation boundary without moving the head."""
    saved, _ = self.right_stack.peek()
    arrived, _ = self.incoming.s["F"].peek()
    from_stack = neg(self.right_stack.empty())
    self.circuit.require(neg(self.incoming.s["F"].empty()),
                         conjunction(neg(from_stack), neg(self.incoming.empty())))
    return Ref.select(from_stack, saved, arrived)

  def reset(self, enabled=TRUE):
    self.focus = Ref.select(enabled, EMPTY, self.focus)
    self.left_stack.clear(enabled)
    self.right_stack.clear(enabled)
    self.incoming.clear(enabled)

  def right(self, enabled=TRUE):
    self.circuit.require(self.can_right(), enabled)
    self.left_stack.push(self.focus, enabled=enabled)
    from_stack = neg(self.right_stack.empty())
    from_queue = conjunction(enabled, neg(from_stack))
    saved, _ = self.right_stack.pop(conjunction(enabled, from_stack))
    self.incoming.work(from_queue)
    arrived = self.incoming.pop(from_queue)
    self.incoming.work(from_queue)
    self.focus = Ref.select(enabled, Ref.select(from_stack, saved, arrived), self.focus)

  def left(self, enabled=TRUE):
    self.circuit.require(neg(self.left_stack.empty()), enabled)
    self.right_stack.push(self.focus, enabled=enabled)
    saved, _ = self.left_stack.pop(enabled)
    self.focus = Ref.select(enabled, saved, self.focus)

  def copy_from(self, other, enabled=TRUE):
    self.focus = Ref.select(enabled, other.focus, self.focus)
    self.left_stack.copy_from(other.left_stack, enabled)
    self.right_stack.copy_from(other.right_stack, enabled)
    self.incoming.copy_from(other.incoming, enabled)

  def finalize(self, maintain=True):
    if maintain:
      self.incoming.work()
    self.left_stack.finalize(); self.right_stack.finalize()
    self.incoming.finalize()
    self.circuit.put_ref(self.focus_key, self.focus)


class PlaceHead:
  def __init__(self, head_pool, queue_pool, counter_pool, name, alphabet="ab"):
    self.head = InputHead(head_pool, queue_pool, counter_pool, name, alphabet)
    self.circuit, self.name = self.head.circuit, name
    self.gap_key = name + ".gap"
    self.gap = self.circuit.get(PREVIOUS, self.gap_key, (False, True), True).eq(True)

  def append(self, cell, enabled=TRUE):
    self.head.append(cell, enabled)

  def read(self):
    value = self.head.read()
    return Value.select(value.eq(None), Value.constant(None),
                        Value.select(self.gap, Value.constant("s"), value))

  def can_right(self):
    return disjunction(neg(self.gap), self.head.can_right())

  def right(self, enabled=TRUE):
    self.head.right(conjunction(enabled, self.gap))
    self.gap = choose(enabled, neg(self.gap), self.gap)

  def left(self, enabled=TRUE):
    self.circuit.require(neg(self.head.read().eq(None)), enabled)
    self.head.left(conjunction(enabled, neg(self.gap)))
    self.gap = choose(enabled, neg(self.gap), self.gap)

  def is_first(self):
    before, _ = self.head.left_stack.peek()
    return conjunction(neg(self.gap), neg(self.head.left_stack.empty()), neg(before.present()))

  def copy_from(self, other, enabled=TRUE):
    self.head.copy_from(other.head, enabled)
    self.gap = choose(enabled, other.gap, self.gap)

  def finalize(self):
    self.head.finalize()
    self.circuit.put(self.gap_key, Value.select(self.gap, Value.constant(True), Value.constant(False)))
